### PR TITLE
[MERGE] mail, crm: allow to log a feedback when marking as lost

### PR DESCRIPTION
**Pull request by tde-banana-odoo**
PURPOSE

Allow sales reps to add a closing note to their lead while it is being marked
as lost. This is currently already done by a lot of sales reps but manually
with the "Log a Note" button.

SPECIFICATIONS

In ``Lost Reason`` model: add a new html field allowing to log a note on the
lost leads. Below the m2o, add an Extra Comment field where users can add a
"closing note". When the wizard is submitted, log this message as a note on
selected records.

Allow to link this feedback to the tracking done when marking as lost to have a
single message. Tracking is currently done by accumulating changes in a
structure (see ``env.cr.precommit.data`` usage with ``mail.tracking.<name>``
key). In the end those values are used to generate a message with tracking
value and a subtype. In this merge we allow to manually set a body used as a
message for the tracking message. This is used in the lost wizard to merge the
tracking and the feedback message.

Correctly name lost_reason field on lead. As this is a many2one, it should end
with an ``_id`` suffix. Otherwise we may think this is a char field, which was
probably the case at one point.

While working on tracking, rename tracking methods and clean dead code.
This should not change anything, just renaming some old code.

Task-2671709

### CODE REVIEW

File: addons/test_mail/tests/test_performance.py [hunk 1]
```diff
@@ -47,6 +47,12 @@ def setUp(self):
         super(BaseMailPerformance, self).setUp()
         # patch registry to simulate a ready environment
         self.patch(self.env.registry, 'ready', True)
+        self.flush_tracking()
+
+    def flush_tracking(self):
+        """ Force the creation of tracking values. """
+        self.env['base'].flush()
+        self.cr.flush()
```
Comment (awa-odoo) [inline addons/test_mail/tests/test_performance.py:55]: Looks a lot like a "flush the whole world to make it work".

Can't we just flush `mail.tracking.value` ?

File: addons/crm/tests/common.py [hunk 2]
```diff
@@ -242,6 +242,10 @@ def setUpClass(cls):
             'res_id': cls.activity_type_1.id,
         })
 
+    def setUp(self):
```
Comment (awa-odoo) [inline addons/crm/tests/common.py:245]: Commit message about the flush seems a bit strange in terms of wording.
Maybe reword a bit and explain exactly why?

> We also specifically add some tracking flush in test performance tests

typo?

File: addons/test_mail/tests/test_performance.py [hunk 3]
```diff
@@ -452,6 +458,8 @@ def test_message_subscribe_subtypes(self):
     @warmup
     def test_message_track(self):
         record = self.env['mail.performance.tracking'].create({'name': 'Zizizatestname'})
+        record.flush()
```
Comment (awa-odoo) [inline addons/test_mail/tests/test_performance.py:461]: Not using the new method?

File: addons/crm/tests/test_crm_lead_lost.py [hunk 4]
```diff
@@ -18,29 +19,96 @@ def setUpClass(cls):
...
+        self.assertEqual(len(self.lead_1.message_ids), 1, 'Should contain creation message')
+        creation_message = self.lead_1.message_ids[0]
+        self.assertEqual(creation_message.subtype_id, self.env.ref('crm.mt_lead_create'))
+
+        # Update responsible as ACLs is "own only" for user_sales_salesman
         self.lead_1.with_user(self.user_sales_manager).write({
             'user_id': self.user_sales_salesman.id,
             'probability': 32,
         })
+        self.flush_tracking()
```
Comment (awa-odoo) [inline addons/crm/tests/test_crm_lead_lost.py:34]: Should this be in the "flush the world" commit?
A bit random that it appears here.

File: addons/crm/tests/test_crm_lead_lost.py [hunk 5]
```diff
@@ -18,29 +19,96 @@ def setUpClass(cls):
...
 
-        lead = self.lead_1.with_user(self.env.user)
+        lead = self.env['crm.lead'].browse(self.lead_1.ids)
```
Comment (awa-odoo) [inline addons/crm/tests/test_crm_lead_lost.py:36]: Re-browsing stuff? Why exactly?

File: addons/crm/tests/test_crm_lead_lost.py [hunk 6]
```diff
@@ -18,29 +19,96 @@ def setUpClass(cls):
...
+        self.assertFalse(lead.lost_reason)
         self.assertEqual(lead.probability, 32)
+        self.assertEqual(len(lead.message_ids), 2, 'Should have tracked new responsible')
+        update_message = self.lead_1.message_ids[0]
```
Comment (awa-odoo) [inline addons/crm/tests/test_crm_lead_lost.py:40]: Using self.lead_1 again instead of lead, lost reviewer is lost.

File: addons/crm/tests/test_crm_lead_lost.py [hunk 7]
```diff
@@ -18,29 +19,96 @@ def setUpClass(cls):
...
+        message. """
+        leads = self._create_leads_batch(lead_type='lead', count=10, probabilities=[10, 20, 30])
+        self.assertEqual(len(leads), 10)
+        self.flush_tracking()
+
+        lost_wizard = self.env['crm.lead.lost'].with_context({
+            'active_ids': leads.ids,
+        }).create({
+            'lost_reason_id': self.lost_reason.id,
+            'lost_feedback': '<p>I cannot find it. It was in my closet and pouf, disappeared.</p>',
```
Comment (awa-odoo) [inline addons/crm/tests/test_crm_lead_lost.py:83]: Not bad, I personally would have gone for "I click and boom!" :smile:

File: addons/crm/tests/test_crm_lead_lost.py [hunk 8]
```diff
@@ -18,29 +19,96 @@ def setUpClass(cls):
...
+        })
+        lost_wizard.action_lost_reason_apply()
+        self.flush_tracking()
+
+        for lead in leads:
```
Comment (awa-odoo) [inline addons/crm/tests/test_crm_lead_lost.py:88]: Not sure exactly why you create 10 leads?
Is it to test the batching aspect?
Is there not already other tests for that?

File: addons/project/models/project.py [hunk 9]
```diff
@@ -1509,7 +1509,7 @@ def _get_depends_tracked_fields(self):
         """ Returns the set of tracked field names for the current model.
         Those fields are the ones tracked in the parent task when using task dependencies.
 
-        See :meth:`mail.models.MailThread._get_tracked_fields`"""
+        See :meth:`mail.models.MailThread._tracking_get_fields`"""
```
Comment (awa-odoo) [inline addons/project/models/project.py:1512]: _track_get_fields

File: addons/mail/models/mail_thread.py [hunk 10]
```diff
@@ -547,21 +547,21 @@ def _message_track_post_template(self, changes):
     def _track_template(self, changes):
         return dict()
 
-    def message_track(self, tracked_fields, initial_values):
+    def _message_track(self, field_names, initial_values):
```
Comment (awa-odoo) [inline addons/mail/models/mail_thread.py:560]: Why not "_track_message" then? Confusing? "_track_values"? "_track_all_the_things"?

File: addons/crm/tests/test_crm_lead_lost.py [hunk 11]
```diff
@@ -92,7 +92,7 @@ def test_lead_lost_batch_wfeedback(self):
             self.assertEqual(lead.probability, 0)
             self.assertEqual(lead.lost_reason_id, self.lost_reason)
             # check messages
-            self.assertEqual(len(lead.message_ids), 3, 'Should have 3 messages: creation, lost, and log')
```
Comment (awa-odoo) [inline addons/crm/tests/test_crm_lead_lost.py]: Commit message says "when losting a lead" ? loosing? or marking as lost if you want to be specific.

Also: 

> from user an UI point of view

typo?

File: addons/crm/wizard/crm_lead_lost.py [hunk 12]
```diff
@@ -18,12 +18,12 @@ class CrmLeadLost(models.TransientModel):
...
         leads = self.env['crm.lead'].browse(self.env.context.get('active_ids'))
-        res = leads.action_set_lost(lost_reason_id=self.lost_reason_id.id)
         if not is_html_empty(self.lost_feedback):
-            leads._message_log_batch(
-                bodies=dict(
-                    (lead.id, self.lost_feedback)
-                    for lead in leads),
-                subject=False,
+            leads._track_set_log_message(
+                '<div style="margin-bottom"><p>%s:</p>%s<br /></div>' % (
```
Comment (awa-odoo) [inline addons/crm/wizard/crm_lead_lost.py:23]: I don't think margin-bottom on its own without value does anything?

File: addons/mail/models/mail_thread.py [hunk 13]
```diff
@@ -569,6 +579,8 @@ def _message_track(self, field_names, initial_values):
             except MissingError:
                 continue
 
+        # find content to log as body
+        bodies = self.env.cr.precommit.data.pop(f'mail.tracking.message.{self._name}', {})
```
Comment (awa-odoo) [inline addons/mail/models/mail_thread.py:583]: So this basically looks like "context usage but without the context", right?
A bit strange and confusing.

**Review by awa-odoo — COMMENTED** [review]
Hello @tde-banana-odoo 

Nice small improvement, with some strange side stuff.
Added a few polishing comments.

Side-note: are you going to write a merge message?

Cheers!

File: addons/test_mail/tests/test_performance.py (hunk 3 above)
Comment (tde-banana-odoo) [reply addons/test_mail/tests/test_performance.py]: Actually not necessary. I will just keep the initial flush_tracking done in setup, other flushes are done by warmup / query counters. I will update commit message accordingly.

File: addons/test_mail/tests/test_performance.py (hunk 1 above)
Comment (tde-banana-odoo) [reply addons/test_mail/tests/test_performance.py]: It is a copy of what is done in mail test class. This was done by the framework I think.

File: addons/test_mail/tests/test_performance.py [hunk 14]
```diff
@@ -47,6 +47,12 @@ def setUp(self):
         super(BaseMailPerformance, self).setUp()
         # patch registry to simulate a ready environment
         self.patch(self.env.registry, 'ready', True)
+        self.flush_tracking()
+
+    def flush_tracking(self):
```
[the remark this answers is not in the document]
Comment (tde-banana-odoo) [reply addons/test_mail/tests/test_performance.py]: I did like mail. But ok to make it private :) .

File: addons/crm/tests/common.py (hunk 2 above)
Comment (tde-banana-odoo) [reply addons/crm/tests/common.py:245]: Typo indeed. Commit message updated.

File: addons/crm/tests/test_crm_lead_lost.py (hunk 4 above)
Comment (tde-banana-odoo) [reply addons/crm/tests/test_crm_lead_lost.py:34]: Well, yes and no. There are more additions to that test... but ok, added the flush + message number part in the previous commit, and re-updated its message. Hopefully it is clearer now :) .

File: addons/crm/tests/test_crm_lead_lost.py (hunk 5 above)
Comment (tde-banana-odoo) [reply addons/crm/tests/test_crm_lead_lost.py:36]: Cache issue often ... better browse than doing with_user (ACLs notably).

File: addons/crm/tests/test_crm_lead_lost.py (hunk 6 above)
Comment (tde-banana-odoo) [reply addons/crm/tests/test_crm_lead_lost.py]: Indeed :D .

File: addons/crm/tests/test_crm_lead_lost.py (hunk 8 above)
Comment (tde-banana-odoo) [reply addons/crm/tests/test_crm_lead_lost.py:88]: Indeed, to test batch and ensure it works as intended with active_ids. There are no other tests for lost wizard as far as I know, so indeed this was not done.

File: addons/mail/models/mail_thread.py (hunk 10 above)
Comment (tde-banana-odoo) [reply addons/mail/models/mail_thread.py]: This is a "main API method". All main methods are message_something.

File: addons/mail/models/mail_thread.py (hunk 13 above)
Comment (tde-banana-odoo) [reply addons/mail/models/mail_thread.py:572]: This is done like the tracking actually. It accumulates data and generate a message in the end. In addition to values, we may now specificy a body.

**Review by tde-banana-odoo — COMMENTED** [review]
Thanks for the riviou, here are the replies !

**Comment by awa-odoo** [comment]
@robodoo r+ rebase-merge
